snuba/query/processors/conditions_enforcer.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from typing import Set
+
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Expression, Query
+from snuba.query.conditions import (
+    FUNCTION_TO_OPERATOR,
+    condition_pattern,
+    get_first_level_and_conditions,
+)
+from snuba.query.expressions import Column
+from snuba.query.matchers import AnyExpression
+from snuba.request.request_settings import RequestSettings
+from snuba.state import get_config
+
+logger = logging.getLogger(__name__)
+
+
+class MandatoryConditionEnforcer(QueryProcessor):
+    """
+    Ensures the query contains conditions on aa specific set of columns
+    as top level conditions.
+
+    This is supposed to be a failsafe mechanism to ensure the query
+    processing pipeline does not drop conditions that are essential for
+    safety (like conditions on project_id or timestamp).
+
+    If this processor fails the query is supposed to fail as in that case
+    it would be better to make a query fail that accidentally over
+    exposing data for missing conditions.
+    """
+
+    def __init__(self, required_columns: Set[str]) -> None:
+        self.__required_columns = required_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        found_cols = set()
+
+        pattern = condition_pattern(
```

Comment:
This pattern is too generic. Every top level condition will match it. More importantly, I think you made it this way so that columns inside complex expressions are still counted as valid mandatory conditions, and that is not true everywhere else in the code. Why would we count columns inside complex expressions as valid mandatory conditions?

Reply:
> Why would we count columns inside complex expressions as valid mandatory conditions?

Because this happens at the end of query processing.
I want to allow `nullIf(toStartOfHour(toDateTime(.....))) > 'something'` to be considered as valid. I am concerned that making this failsafe logic complex enough to perfectly match all the cases we want for valid conditions would make the logic itself so complex that we could break it easily.

Anyway I have a better idea: I will only deal with project_id as that is the critical field and ignore time conditions. Those are less of a problem.